src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/RuleDriver.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.s
+
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    public interface IRuleDriverFactory
+    {
+        IRuleDriver GetRuleDriver();
+    }
+
+    public class RuleDriverFactory : IRuleDriverFactory
+    {
+        private RuleDriver _driver;
+        public IRuleDriver GetRuleDriver()
+        {
+            if (_driver == null)
+                _driver = new RuleDriver();
+
+            return _driver;
+        }
+    }
+
+    public interface IRuleDriver
+    {
+        IEnumerable<CompatDifference> Run<T>(ElementMapper<T> mapper);
+    }
+
+    internal class RuleDriver : IRuleDriver
+    {
+        private readonly IEnumerable<Rule> _rules;
+
+        internal RuleDriver()
+        {
+            _rules = GetRules();
+        }
+
+        public IEnumerable<CompatDifference> Run<T>(ElementMapper<T> mapper)
+        {
+            List<CompatDifference> differences = new();
+            if (mapper is AssemblyMapper am)
+            {
+                Run(am, differences);
+            }
+            if (mapper is TypeMapper tm)
+            {
+                Run(tm, differences);
+            }
+            if (mapper is MemberMapper mm)
+            {
+                Run(mm, differences);
+            }
+
+            return differences;
+        }
+
+        private void Run(AssemblyMapper mapper, List<CompatDifference> differences)
+        {
+            foreach (Rule rule in _rules)
+            {
+                rule.Run(mapper, differences);
+            }
+        }
+
+        private void Run(TypeMapper mapper, List<CompatDifference> differences)
+        {
+            foreach (Rule rule in _rules)
+            {
+                rule.Run(mapper, differences);
+            }
+        }
+
+        private void Run(MemberMapper mapper, List<CompatDifference> differences)
+        {
+            foreach (Rule rule in _rules)
```

Comment:
This seems like it might have some overhead when we have lots of rules.   I wonder if there is a better pattern to use here?  Might be something to think about / measure.  Events/handlers seem like a similar problem, though they are a bit harder to use 🤔

Reply:
Yes, this is something I would like to revisit afterwards and improve. Maybe we can have rules hook up to specific events, like MemberMapperRuleEvent, so that we don't iterate on all the rules for a MemberMapper and just the rules that need to run on it?

Reply:
I chatted with @ericstj offline, I will open a follow up issue so that we can make progress on this but this is something that I should investigate.